Retrofit 설정 및 로그인 API 구현

Kim-DG opened the pull request:
- [x] Retrofit 설정 완료
- [x] 로그인 API 구현 및 테스트 완료
- [x] authenticate 구현
- [ ] authenticate 테스트

seunggom committed:
ingame domain, data 영역 추가 / 폴더 구조 변경

Kim-DG committed:
[Update] retrofit 세팅, 로그인 api 적용

seunggom commented on data/src/main/java/com/keunsori/data/datasource/LocalDataSource.kt:
DataStore에 값 읽고 쓸 때에 예외처리 하면 좋을 것 같습니다.

그리고 DataStore 쓰는 것은 좋은데, 사용하는 방식이 잘못된 것 같아요.
DataStore로부터 값을 읽으면 `Flow` 를 반환하는데, 지금 코드를 보면 전혀 Flow, coroutine을 활용하지 않고 refreshToken 값을 refreshToken.first() 를 사용하여 값을 얻어오려고 하고 있는 것이 문제인 것 같습니다.
그래서 RefreshToken이 필요한 지점에서 저 Flow<String>을 구독하여 값을 받아오도록 하든지, 아니면 LocalDataSource 초기화 시점에 최초 한 번만 dataStore로부터 refreshToken 값을 읽어와서 캐싱을 해두든지 하면 좋을 것 같습니다.

Kim-DG replied:
refreshToken은 지속적으로 데이터를 받아올 필요가 없을 것 같아 collect를 사용하여 구독하지 않고 필요한 시점에 한번만 가져와도 될 것 같아 first()를 사용했는데 혹시 이러한 경우에도 Flow를 활용하는게 좋을까요? 지금은 후자의 방법으로 수정 하였습니다.

https://github.com/Keunsori-Developer/towerdle-client/pull/8/files#diff-6f7143e499ba6e24a037d81dbf57fcde51e4f66c258392ad672fe94d76ad7006R35-R46

seunggom replied:
first()도 가능하긴 한데, 호출 시점에 따라 수집된 데이터가 하나도 없으면 exception이 발생할 수 있을 것 같습니다.

저 같은 경우엔 데이터를 1개만 수집하고자 할 때에는 flow.**take(1)**.collect {} 이런식으로 사용했습니닷

seunggom commented on data/src/main/java/com/keunsori/data/datasource/UserRemoteDataSource.kt:
```diff
@@ -1,14 +1,27 @@
 class UserRemoteDataSource @Inject constructor(
-    private val apiService: ApiService,
+    private val apiService: RetrofitService,
+    private val oauthApiService: OauthRetrofitService
```
`RetrofitService`와 `OauthRetrofitService`는 무슨 차이 인가요?

그리고 `UserRemoteDataSource` 가 `OauthRetrofitService` 객체를 가지고 있을 필요는 없고, 그 안에 구현되어 있는 OauthApiService 객체를 가지고 있으면 될 것 같습니다.

마침 Hilt 사용하고 있으니, https://jtm0609.tistory.com/22 요고 참고해서 모듈로 빼서 주입되도록 하면 좋을 것 같아요.

Kim-DG replied:
인증을 위한 Interceptor가 들어간 RetrofitService와 구분하려 따로 구현하였는데 RetrofitService에 두개의 service를 구현하는게 나을까요?

seunggom replied:
아하 RetrofitService 안에서 두 개의 Service를 관리하는 것도 좋아보입니다요

seunggom replied:
어차피 NetworkModule 하나 만들면 되기 땜에 RetrofitService 클래스도 굳이 필요 없을 것 같습니다

seunggom commented on data/src/main/java/com/keunsori/data/api/OauthApiService.kt:
```diff
@@ -0,0 +1,16 @@
+import retrofit2.http.Body
+import retrofit2.http.POST
+
+interface OauthApiService {
```
```suggestion
interface AuthApiService {
```

Oauth는 사용자 인증과 관련된 프로토콜 이름이라, 요기에는 안맞는 이름인 것 같습니다.

seunggom commented on data/src/main/java/com/keunsori/data/retrofit/RetrofitService.kt:
```diff
@@ -0,0 +1,94 @@
+        Log.i("Authenticator", response.toString())
+        Log.i("Authenticator", "토큰 재발급 시도")
+        return try {
+            runBlocking {
```
runBlocking 을 사용하면 UI Thread 가 Blocking 될텐데, 토큰 갱신하는 로직이 오래걸리게 되면 ANR 이 발생할 수 있어 안전한 방식이 아닌 것 같습니다.

Kim-DG replied:
수정했는데 조금 코드가 복잡한거 같아서 확인 한번 부탁드립니다

https://github.com/Keunsori-Developer/towerdle-client/pull/8/files#diff-34f120cb7951fc6e67750baffcd26f47416f48394c79b8d4feb4440b4a95f236R46-R68

seunggom commented on data/src/main/java/com/keunsori/data/retrofit/RetrofitExtension.kt:
```diff
@@ -0,0 +1,35 @@
+            if (response.isSuccessful) {
+                Log.d("HttpResponse", "success: ${response.code()} ${response.message()}")
+                Log.d("HttpResponse", "response: $res")
+                it.resumeWith(Result.success(res!!)) // respnose가 success일 때 body가 null이면 exception 발생하게 유도
```
res 가 null 이면 Result.failure 를 반환하면 되는 거 아닌가요? 요 코드의 의도가 궁금합니다.

Kim-DG replied:
Result.success라도 res가 null이 아니란걸 보장할 수 없어 위의 방식으로 구현하였습니다.
다시 생각해보니 에러를 발생시키는 것 보다 분기 조건을 더 추가하면 좋을 것 같아 수정하였습니다.

https://github.com/Keunsori-Developer/towerdle-client/pull/8/files#diff-9fd3ff093955e0844f0c40e797f7b534715a10bfd0f30a4ebb11004d6f1ecb02R17-R26

Kim-DG committed:
Update data/src/main/java/com/keunsori/data/api/OauthApiService.kt

Co-authored-by: SeungYeon Kim <47077179+seunggom@users.noreply.github.com>

Kim-DG committed:
[Refactoring] 코드 리뷰 리팩토링

- RetrofitService 모듈화
- TokenAuthenticator runblocking 제거
- refresh token 캐싱

Kim-DG committed:
[Refactoring] MVI 코드 수정 및 게스트 로그인 여부 추가

- view에서 viewmodel의 sendEvent 메소드 만으로 컨트롤할 수 있게 수정
- localdatasource에 게스트로그인여부 추가

seunggom commented:
dev에 머지해주시면 rebase 해서 작업하겠슴다
